### PR TITLE
Bugfix scatter_reduce backward formulas

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76523

Fixes:
1) Backward formula of reduce=prod (previously this was `(grad * result / input).masked_fill_(input == 0, 0)` as adapted from [pytorch_scatter](https://github.com/rusty1s/pytorch_scatter/blob/master/csrc/scatter.cpp#L95-L106) but this is buggy when the input is 0). For that case we need to calculate grad * prod(elements scattered to that index excluding the element we're calculating the gradient for).

2) Backward formula of reduce=amax/amin - when there are multiple elements scattered to an index that correspond to the max/min the gradient should be evenly distributed as in `evenly_distribute_backward`

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -5370,10 +5370,17 @@ std::tuple<Tensor, Tensor> scatter_reduce_backward(
     grad_self = grad;
     grad_src = grad.gather(dim, index);
   } else if (reduce == "prod") {
-    grad_self = (grad * result) / self;
-    grad_self.masked_fill_(self == 0, 0);
-    grad_src = (grad * result).gather(dim, index) / src;
-    grad_src.masked_fill_(src == 0, 0);
+    // Explicitly compute exclusive prod for elements in self/src that are 0
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:5373]: Could we re-use the logic from `prod_backward` (in this same file) here?

File: test/test_autograd.py [hunk 2]
```diff
@@ -7066,6 +7066,30 @@ def test_min_max_median_backprops_to_all_values(self, device):
                 self.assertEqual(x.grad.sum(), 1.)
                 self.assertEqual((x.grad == 1 / 3).sum(), 3)
 
+    def test_scatter_reduce_backprop_edge_cases(self, device):
```
Comment (albanD) [inline test/test_autograd.py:7069]: Doesn't scatter reduce has an OpInfo? These should be test samples there?

File: test/test_autograd.py [hunk 3]
```diff
@@ -7066,6 +7066,30 @@ def test_min_max_median_backprops_to_all_values(self, device):
...
+        # test that gradients are propagated correctly when zeros in self/src are reduced for reduce=prod
+        input = torch.tensor([[0., 1., 2.], [3., 4., 0.]], device=device, requires_grad=True)
+        src = torch.tensor([[1., 2., 3.], [4., 5., 0.]], device=device, requires_grad=True)
+        idx = torch.tensor([[0, 0, 0], [1, 1, 1]], device=device)
+        expected_input_grad = torch.tensor([[6., 1., 1.], [1., 0., 1.]])
+        expected_src_grad = torch.tensor([[0., 0., 0.], [0., 0., 80.]])
+        out = input.scatter_reduce(1, idx, src, 'prod', include_self=True)
+        out.sum().backward()
+        self.assertEqual(input.grad, expected_input_grad)
```
Comment (albanD) [inline test/test_autograd.py:7078]: Any reason not to use gradcheck here?

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (mikaylagawarecki) [reply torch/csrc/autograd/FunctionsManual.cpp:5373]: Hm, I'm not quite sure how we would use cumprod here to calculate the exclusive product given that there's a scatter operation.

File: test/test_autograd.py [hunk 4]
```diff
@@ -7066,6 +7066,30 @@ def test_min_max_median_backprops_to_all_values(self, device):
                 self.assertEqual(x.grad.sum(), 1.)
                 self.assertEqual((x.grad == 1 / 3).sum(), 3)
 
+    def test_scatter_reduce_backprop_edge_cases(self, device):
+        # test that gradients are propagated correctly when zeros in self/src are reduced for reduce=prod
+        input = torch.tensor([[0., 1., 2.], [3., 4., 0.]], device=device, requires_grad=True)
```
[the remark this answers is not in the document]
Comment (mikaylagawarecki) [reply test/test_autograd.py:7071]: You're right, thanks!

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 1 above)
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:5373]: It might be harder to re-use that in the scatter case indeed because you don't have a `cumprod` version of scatter reduce?
It was just a high level question to know if we can de-dupe code there. Also the cumprod version makes it easy to ignore the problems with 0s :p

File: torch/testing/_internal/common_methods_invocations.py [hunk 5]
```diff
@@ -7067,6 +7067,41 @@ def _gather(shape, index_dim, max_indices):
...
+        # This sample tests gradients for the following cases
+        # (a) 1 zero reduced (from src (self[0, 1]), from self (self[0, 0]))
+        # (b) 2 zeros reduced (1 from src and 1 from self (self[1, 0]), 2 from src (self[1, 1]))
+        # (c) > 2 zeros reduced (2 from src, 1 from self (self[2, 0]))
+        # (d) no zeros reduced (self([2, 1]))
+        input = torch.tensor([[0, 13], [0, 17], [0, 19]], dtype=dtype, device=device, requires_grad=requires_grad)
+        src = torch.tensor([[0, 1, 2, 3], [0, 4, 0, 0], [0, 0, 5, 6]], dtype=dtype, device=device, requires_grad=requires_grad)
+        idx = torch.tensor([[1, 1, 0, 0], [0, 0, 1, 1], [0, 0, 0, 1]], dtype=torch.long, device=device)
+
+        sample_inputs.append(
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7082]: fyi the sample inputs functions can be actual generators now and you can just yield SampleInputs as you create them instead of collecting them and return them in an iterable

File: torch/testing/_internal/common_methods_invocations.py [hunk 6]
```diff
@@ -7067,6 +7067,41 @@ def _gather(shape, index_dim, max_indices):
             )
         )
 
+    # Sample inputs to test edge cases for backward
+    # Check that gradients are propagated correctly for prod when zeros in self/src are reduced
+    if requires_grad and reduce == 'prod':
```
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:7072]: this might be more cleanly modeled as a separate function for prod, that first yields from the common sample inputs for this class of ops and then extends the samples like this, but nbd

File: test/test_autograd.py [hunk 7]
```diff
@@ -7066,6 +7066,20 @@ def test_min_max_median_backprops_to_all_values(self, device):
...
+    def test_scatter_reduce_prod_gradgrad_error(self, device):
+        # test that double backward raises an error
+        input = torch.tensor([1.], device=device, dtype=torch.float64, requires_grad=True)
+        src = torch.tensor([0., 0.], device=device, dtype=torch.float64, requires_grad=True)
+        idx = torch.tensor([0, 0], device=device, dtype=torch.long)
+
+        def fn(input, src):
+            return input.scatter_reduce(0, idx, src, 'prod')
+
+        # check that gradcheck doesn't raise an error
```
Comment (albanD) [inline test/test_autograd.py:7078]: nit: we actually want to make sure that gradcheck passes here are it is the only place where we check it for this case!

File: torch/testing/_internal/common_methods_invocations.py [hunk 8]
```diff
@@ -17500,6 +17520,8 @@ def generate_std_var_kwargs(t: torch.Tensor, **kwargs):
         # and scatter_reduce hasn't been added to the whitelist in gen_variable_type yet
         dtypes=all_types_and(torch.float16, torch.bfloat16, torch.bool),
         dtypesIfCUDA=floating_types_and(torch.float16, torch.bfloat16),
+        # FIXME: vmap segfaults in _remove_batch_dim after node logic was added to scatter_reduce_backward(prod)
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:17523]: Ouch
cc @zou3519

File: torch/testing/_internal/common_methods_invocations.py [hunk 9]
```diff
@@ -7578,18 +7578,38 @@ def _gather(shape, index_dim, max_indices):
...
 
-    return sample_inputs
+
+    # Sample inputs to test edge cases for backward
+    # Check that gradients are propagated correctly for prod when zeros in self/src are reduced
+    if requires_grad and reduce == 'prod':
+        # This sample tests gradients for the following cases
+        # (a) 1 zero reduced (from src (self[0, 1], self[1, 1]), from self (self[0, 0], self[2, 0]))
+        # (b) 2 zeros reduced (1 from src and 1 from self (self[1, 0])
+        # (c) no zeros reduced (self([2, 1]))
```
Comment (albanD) [inline torch/testing/_internal/common_methods_invocations.py:7594]: nit: I think it could be good to mention here that the case with 2 zeros in src is tested separately in test_autograd.

**Review by albanD — APPROVED** [review]
SGTM!

**Review by albanD — APPROVED** [review]
SGTM thanks for the update!

File: torch/testing/_internal/common_methods_invocations.py (hunk 8 above)
Comment (mikaylagawarecki) [reply torch/testing/_internal/common_methods_invocations.py:17525]: This segfault might have been due to a print statement I had added somewhere locally, my bad. I don't think there was a segfault otherwise, will remove this.

Comment (mikaylagawarecki) [reply torch/testing/_internal/common_methods_invocations.py:17525]: My bad, this was a mistake on my part likely introduced by a print statement.
